### ISSUE TITLE
Events are not correlated in Application Insights

### ISSUE THREAD

**Issue by fgheysels** [arcus-azure/arcus.observability#343]
**Describe the bug**
I have a simple program which logs a servicebus request to Application Insights, and also logs an Event to Application insights using the ILogger extensions that Arcus provides.
Simplified, it looks like this:

```csharp
using (_logger.BeginScope(logProperties))
{
    var duration = DurationMeasurement.Start();
    bool successFullRequest = true;

    try
    {
          _logger.LogEvent("DataSync message received", logProperties);
          // do some logic
    }
    catch (Exception ex)
    {
        successFullRequest = false;

        _logger.LogCritical(ex, "An error occured while ....");

        throw;
    }
    finally
    {
        _logger.LogServiceBusQueueRequest(_serviceBusNamespace, _queueName, "DataSyncRouteEnvelope", successFullRequest, duration);
    }
```

When I look in App Insights, I can see that the request is logged, but the Event is not found in the timeline or in the logged telemetry that is correlated to that request.   I would expect that I can see that event here as well:

![image](https://user-images.githubusercontent.com/3605786/155949359-03001a88-ee25-488a-a416-e12582db0351.png)

![image](https://user-images.githubusercontent.com/3605786/155949420-ae01eec8-db89-46d0-a891-5a7c5c545165.png)

The event is logged however, as I can find it back.  It is just not correlated to anything:

![image](https://user-images.githubusercontent.com/3605786/155949504-1887d870-4479-4686-8767-25b363b2ec5a.png)

![image](https://user-images.githubusercontent.com/3605786/155949536-c5d53bc7-08d6-4311-a353-32b8c6f96818.png)

**Additional context**
 - Version: Arcus.Observability.Telemetry.Serilog.Sinks.ApplicationInsights 2.4.0
 - .NET Core 3.1

**Comment by stijnmoreels** [comment]
Hmm, but to what should it be correlated? To the request? I think this is maybe an extension to the service-to-service correlation functionality that we're working on, but this is by all means not yet finished. Do we actually say that we already support this?

**Comment by stijnmoreels** [comment]
Ah, could this also be bc custom events are not having a start/duration time?

**Comment by stijnmoreels** [comment]
> When you log a trace (using f.i. `LogInformation`) that trace is also correlated to a parent. It should work in the same way.

Ah, and what is the correlation information model at this point? Is the operation parent ID filled-out? Bc this value will be added to the parent ID of the trace.

**Comment by stijnmoreels** [comment]
Any update on this?

**Comment by stijnmoreels** [comment]
@fgheysels Any update on this?

**Comment by fgheysels** [comment]
No sorry, didn't have the time yet to look at this further.

**Comment by stijnmoreels** [comment]
> No sorry, didn't have the time yet to look at this further.

I'll try to reproduce it.

**Comment by stijnmoreels** [comment]
Ok, I've created the most minimal application with a hosted service that has the same logging structure that you presented in the issue description. When looking at the Application Insights portal, you see that operation ID set, just like the transaction ID and the operation parent ID.

![image](https://user-images.githubusercontent.com/9039753/169506287-2ee1d348-009d-4eaf-b0b7-2116fb1a0330.png)

What the application contains:
### Serilog configuration
```csharp
 logConfig.MinimumLevel.Debug()
          .MinimumLevel.Override("Microsoft", LogEventLevel.Information)
          .Enrich.FromLogContext()
          .Enrich.WithComponentName("Azure HTTP Trigger")
          .Enrich.WithVersion()
          .Enrich.WithCorrelationInfo(serviceProvider)
          .WriteTo.Console()
          .WriteTo.AzureApplicationInsights("<key>");
```
Note that both the Azure Application Insights Serilog sink is configured, as well as the correlation enricher which takes in the `IServceProvider` so it has access to the correlation accessors of the application.

### Application code
```csharp
 public class MyService : IHostedService
    {
        private readonly ILogger<MyService> _logger;

        /// <summary>
        /// Initializes a new instance of the <see cref="MyService" /> class.
        /// </summary>
        public MyService(
            ICorrelationInfoAccessor accessor,
            ILogger<MyService> logger)
        {
            _logger = logger;
            accessor.SetCorrelationInfo(new CorrelationInfo("operation-id", "transaction-id"));
        }

        /// <summary>
        /// Triggered when the application host is ready to start the service.
        /// </summary>
        /// <param name="cancellationToken">Indicates that the start process has been aborted.</param>
        public async Task StartAsync(CancellationToken cancellationToken)
        {
            var logProperties = new Dictionary<string, object>
            {
                ["EventCorrelation"] = "Tests"
            };

            using (_logger.BeginScope(logProperties))
            {
                using (var duration = DurationMeasurement.Start())
                {
                    bool successFullRequest = true;
                    try
                    {
                        _logger.LogEvent("DataSync message received", logProperties);
                        await Task.Delay(TimeSpan.FromSeconds(1), cancellationToken);
                    }
                    catch (Exception ex)
                    {
                        successFullRequest = false;

                        _logger.LogCritical(ex, "An error occured while ....");

                        throw;
                    }
                    finally
                    {
                        _logger.LogServiceBusQueueRequest("servicebus namespace", "queue name", "DataSyncRouteEnvelope", successFullRequest, duration);
                    }
                }
            }
        }
```
Note that the correlation of the request being logged here should be set before the actual correlated telemetry. In Web API, we have done this via the HTTP correlation middleware component, in the messaging library this is done via the Azure Service Bus message router that does this before the actual processing of the message handling.

In conclusion, the functionality here works as expected and the events are correlated with the correlation information from the accessors.

Could you provide any more information on what parts you struggle? Thx.

**Comment by stijnmoreels** [comment]
If the cause of the problem is that the correlation was not yet set correctly before writing telemetry, then we should maybe mention it even more with danger signs and alarm bells, as it's already in bold capitals now 😄 .

![image](https://user-images.githubusercontent.com/9039753/169792885-abef181a-21e7-49d1-9936-b5ae797299f3.png)

**Comment by fgheysels** [comment]
I see that I'm missing the `WithCorrelationInfo` enricher when setting up logging.
However, I need to pass in a `ServiceProvider` , but how can I get one ?   It looks like at the moment where I'm configuring logging, there is no ServiceProvider available yet.

This is how I set it up (left out some additional code, so only the basics here):

```csharp
Host.CreateDefaultBuilder(args)
    .ConfigureAppConfiguration((context, config) =>
    {
    })
    .UseSerilog(ConfigureLogger)
    .ConfigureWebHostDefaults(webBuilder =>
    {
        webBuilder.ConfigureKestrel(kestrelServerOptions => kestrelServerOptions.AddServerHeader = false)
                      .UseUrls(httpEndpointUrl)
                      .ConfigureLogging(logging => logging.AddConsole())
                      .UseStartup<Startup>();
    });
return webHostBuilder;
```

The `ConfigureLogger` method looks like this:

```csharp
private static void ConfigureLogger(HostBuilderContext context, LoggerConfiguration loggerConfiguration)
{
    loggerConfiguration.ReadFrom.Configuration(context.Configuration)
                        .Enrich.FromLogContext()
                        .WriteTo.Console();

    var appInsightsInstrumentationKey = context.Configuration["ApplicationInsights:InstrumentationKey"];

    if (!String.IsNullOrWhiteSpace(appInsightsInstrumentationKey))
    {
        loggerConfiguration.WriteTo.AzureApplicationInsights(appInsightsInstrumentationKey);
    }
}

```

**Comment by fgheysels** [comment]
Ok, I decided to have a look in Arcus.Templates to see how it is done there.   Looks like there's an overload of `UseSerilog` that accepts a delegate that defines a `IServiceProvider`

**Comment by stijnmoreels** [comment]
So we can close this, yes?

**Comment by stijnmoreels** [comment]
Closing this as the general question here is solved.
